text/0000-secret-types.md

```diff
@@ -0,0 +1,206 @@
+long lived
+ - Extremely valuable if compromised
+ - Each bit compromised provides incremental value
+ - Confidentiality of compromise is desirable
+Therefore, it’s important to use data-invariant programming for secrets. For this reason, secret types would only allow data invariant operations at all levels of compilation.
+
+Additionally, secret types serve as an indicator to programmers that this information should be treated with care, and not mixed with non-secret data, an invariant that would be enforced by the type system.
+
+[1] *Cryptographic Software in a Post-Spectre World.* Chandler Carruth. RWC 2020. Recording forthcoming.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Secret integer types are a type of restricted integer primitive. In particular, non-constant time operations like division are prohibited. Printing of secret integers directly is also prohibited--they must first be declassified into non-secret integers.
+
+These integers are intended to form a basis for the creation of other secret types, which can be built on top of them, such as secret strings for storing passwords in them.
+
+Comparison of secret integer types is allowed, but the algorithm is independent of the secrets. Comparison of secret types must return a secret_bool, which cannot be branched on.
+
+Example: comparison
+```
+let x : secret_i8 = 6;
+Let y : secret_i8 = 10;
+if (x < y) { // compiler error: cannot branch on secret bool
+  ...
+}
+```
+Example: declassification
+```
+let x : secret_i8 = 6;
+let y : secret_i8 = 10;
+println!((x ^ y).declassify())
+```
+
+Since indexing vectors and arrays is only possible using `usize,` you will be prevented from indexing into a vector or array with a secret integer. Error messages will derive naturally from the type system.
+
+
+Secret types will likely be a more advanced topic when teaching Rust.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+For each fixed-size integer type:
+Implement the following methods:
+- From_be
+- From_le
+- From_be_bytes
+- From_le_bytes
+- From_ne_bytes
+- Is_positive
+- Is_negative
+- Leading_zeros
+- Min_value
+- Max_value
+- Overflowing_add
+- Overflowing_sub
+- Overflowing_mul
+- Overflowing_neg
+- Overflowing_shl
+- Overflowiing_shr
+- Overflowing_pow
+- Pow
+- Reverse_bits
+- Rotate_left
+- Rotate_right
+- Saturating_add
+- Saturating_sub
+- Saturating_neg
+- Saturating_mul
+- Saturating_pow
+- Signum
+- Swap_bytes
+- To_be
+- To_le
+- To_be_bytes
+- To_le_bytes
+- To_ne_bytes
+- Trailing_zeros
+- Wrapping_add
+- Wrapping_sub
+- Wrapping_mul
+- Wrapping_neg
+- Wrapping_shl
+- Wrapping_shr
+- Wrapping_pow
+
+
+Implement the following traits
+Secret integers may only  be combined with other secret integers and the result will be a secret type
+- Add
+- AddAssign
+- BitAnd
+- BitAndAssign
+- BitOr
+- BitOrAssign
+- BitXor
+- BitXorAssign
+- Clone
+- Copy
+- Default
+- Drop
+- Mul
+- MulAssign
```

Comment:
This is often not constant time on the hardware level, right?

Reply:
* and / or / xor should be constant time
* copy has no machine instructions it's just semantic
* clone is just copy for copy types
* Default just pulls pulls out a 0 for integers
* Drop does nothing probably, or it could volatile write 0 or something, which is still constant time.

Add and Mul I don't know enough to say in all cases.

Reply:
I should have said that I was talking about `Mul` and `MulAssign`.

Reply:
Mul,, and thus MulAssign, are not constant-time on ARM Cortex or Power/PowerPC. Many implementations use smaller multipliers in HW and then stitch the partial products together over multiple clock cycles.  Most such implementations allow some of the clock cycles to be skipped for certain data.  The most common is early termination if either input is zero. The next most common early termination case is if the upper half of either input is zero.  The ARM Cortex family multipliers have an even more complex scheme; it appears to be undocumented, but testing indicates that the Cortex multipliers early terminate whenever the upper half of either input has a Hamming weight less than 2 or more than 14.

Reply:
Many architectures have a constant-time means of converting a status bit (usually carry) to an all-zero or all-one mask. On those architectures conditional if/then/else logic can be implemented by masked xor/masked-and/xor selection to run in constant time. The LLVM IR has the needed primitive, but LLVM usually realizes that primitive as a varying-execution-time conditional branch. That conversion would need to be suppressed for conditional expressions involving these secret types.

Reply:
The constant-time behaviour of multiplication can indeed be a problem on some CPUs, although it's very often safe. Much more information is at https://www.bearssl.org/ctmul.html

However, multiplication is extremely useful! Omitting it at one layer likely means that people will simulate it one layer up. (I.e. people will implement multiplication in Rust code using addition etc.) LLVM knows the CPU target, can know the behaviour of mul on that target, and is likely best placed to substitute a constant-time replacement when needed.

Reply:
I think if you're using a CPU which e.g. short-circuits on multiply-by-zero or multiply-by-one, it's basically broken for cryptographic purposes (e.g. the aforementioned PPC32)

I have no expectation of Rust or LLVM to try to autodetect these CPUs or some sort of countermeasure, but would rather explicitly call out CPU-level assumptions about how e.g. multiplication must work, hopefully as a standard set of boilerplate which can be copied-and-pasted into Rust libraries which rely on those set of assumptions.